File: python/easy/sorted_squared_array/sortedSquaredArray.py
# O(nlog(n)) time | O(n) space
# def sortedSquaredArray(array):
#     sortedArray = [0 for _ in array]
#     for idx in range(len(array)):
#         value = array[idx]
#         sortedArray[idx] = value*value
#     sortedArray.sort()
#     return sortedArray

# O(n) time | O(n) space
def sortedSquaredArray(array):
    sortedSquares = [0 for _ in array]
    leftIdx = 0
    rightIdx = len(array) - 1

    while leftIdx <= rightIdx:
        leftValue = array[leftIdx]
        rightValue = array[rightIdx]
        if abs(leftValue) < abs(rightValue):
            sortedSquares[rightIdx-leftIdx] = rightValue * rightValue
            rightIdx -= 1
        else:
            sortedSquares[rightIdx-leftIdx] = leftValue * leftValue
            leftIdx += 1
    return sortedSquares


print(sortedSquaredArray([-7, -6, -5, -4, -2, 0, 0, 1, 2, 3, 5, 6]))

File: javascript/easy/sorted_squared_array/sortedSquaredArray.js
// O(nlog(n)) time | O(n) space
// function sortedSquaredArray(array) {
//     const sortedArray = new Array(array.length).fill(0);
//     for (let i = 0; i < array.length; i++) {
//         let value = array[i];
//         sortedArray[i] = value * value;
//     }
//     sortedArray.sort((a, b) => a - b);
//     return sortedArray;
// }

function sortedSquaredArray(array) {
    const sortedArray = new Array(array.length).fill(0);
    let leftIdx = 0;
    let rightIdx = array.length - 1;
    while (leftIdx <= rightIdx) {
        let leftValue = array[leftIdx];
        let rightValue = array[rightIdx];
        if (Math.abs(leftValue) < Math.abs(rightValue)) {
            sortedArray[rightIdx - leftIdx] = rightValue * rightValue;
            rightIdx--;
        } else {
            sortedArray[rightIdx - leftIdx] = leftValue * leftValue;
            leftIdx++;
        }
    }
    return sortedArray;
}

console.log(sortedSquaredArray([-7, -6, -5, -4, -2, 1, 2, 3, 5, 6]));